File: iOS/ShengwangBeautyView/Classes/UI/Builders/MakeupPageBuilder.swift
//
//  MakeupPageBuilder.swift
//  BeautyView
//
//  Created by HeZhengQing on 2026/1/22.
//

import Foundation
import UIKit

/// Style makeup module page builder
/// Responsible for building page information for style makeup module
internal class MakeupPageBuilder: IPageBuilder {
    
    private let beautyConfig: ShengwangBeautySDK.BeautyConfig
    
    init(beautyConfig: ShengwangBeautySDK.BeautyConfig) {
        self.beautyConfig = beautyConfig
    }
    
    func buildPage() -> BeautyPageInfo {
        var makeupItems: [BeautyItemInfo] = []
        
        // None effect item
        makeupItems.append(
            BeautyItemInfo(
                name: "beauty_effect_none",
                icon: UIImage.beautyIcon(named: "beauty_ic_none"),
                isSelected: beautyConfig.makeupName == nil,
                showSlider: false,
                type: .none,
                onItemClick: { [weak self] _ in
                    self?.beautyConfig.makeupName = nil
                }
            )
        )
        
        // Style makeup options (only templates present in the material package)
        // Makeup-Young ✓
        addMakeupItem(&makeupItems, name: "beauty_makeup_young", icon: UIImage.beautyIcon(named: "beauty_ic_makeup_young"), makeupName: MakeupNames.young)
        // Makeup-Mature — not in material package, commented out
//        addMakeupItem(&makeupItems, name: "beauty_makeup_mature", icon: UIImage.beautyIcon(named: "beauty_ic_makeup_mature"), makeupName: MakeupNames.mature)
        // Makeup-Aura — not in material package, commented out
//        addMakeupItem(&makeupItems, name: "beauty_makeup_aura", icon: UIImage.beautyIcon(named: "beauty_ic_makeup_aura"), makeupName: MakeupNames.aura)
        // Makeup-Natural ✓
        addMakeupItem(&makeupItems, name: "beauty_makeup_natural", icon: UIImage.beautyIcon(named: "beauty_ic_makeup_natural"), makeupName: MakeupNames.natural)
        // Makeup-Graceful — not in material package, commented out
//        addMakeupItem(&makeupItems, name: "beauty_makeup_graceful", icon: UIImage.beautyIcon(named: "beauty_ic_makeup_graceful"), makeupName: MakeupNames.graceful)
        // Makeup-Charm — not in material package, commented out
//        addMakeupItem(&makeupItems, name: "beauty_makeup_charm", icon: UIImage.beautyIcon(named: "beauty_ic_makeup_charm"), makeupName: MakeupNames.charm)
        // Makeup-Perkey ✓
        addMakeupItem(&makeupItems, name: "beauty_makeup_perky", icon: UIImage.beautyIcon(named: "beauty_ic_makeup_perky"), makeupName: MakeupNames.perky)
        // Makeup-Maiden — not in material package, commented out
//        addMakeupItem(&makeupItems, name: "beauty_makeup_maiden", icon: UIImage.beautyIcon(named: "beauty_ic_makeup_maiden"), makeupName: MakeupNames.maiden)
        // Makeup-Insight — not in material package, commented out
//        addMakeupItem(&makeupItems, name: "beauty_makeup_insight", icon: UIImage.beautyIcon(named: "beauty_ic_makeup_insight"), makeupName: MakeupNames.insight)
        // Makeup-Misty — not in material package, commented out
//        addMakeupItem(&makeupItems, name: "beauty_makeup_misty", icon: UIImage.beautyIcon(named: "beauty_ic_makeup_misty"), makeupName: MakeupNames.misty)
        
        return BeautyPageInfo(
            name: "beauty_group_makeup",
            itemList: makeupItems,
            type: .styleMakeup
        )
    }
    
    /// Add makeup option
    ///
    /// 1. On initialization:
    ///    - If currently selected is this template, use beautyConfig.makeupIntensity (will read from cache)
    ///    - If currently selected is not this template, read intensity value from cache for this template (if not found, display 0f)
    /// 2. When user drags slider: Update current template's intensity value (will sync update cache)
    /// 3. When user clicks to switch template:
    ///    - First get cached intensity value for this makeup (if user set it before)
    ///    - Switch makeup (makeupName), this will trigger SDK to set default intensity
    ///    - If cache has intensity value for this makeup (whether 0 or not), override SDK's default with cached value
    ///    - Update UI displayed intensity value to ensure UI matches SDK state
    ///
    /// Benefits of this design:
    /// - Each template's intensity value will be cached, maintaining user-set intensity when switching back
    /// - API layer manages cache uniformly, UI layer logic is simple and clear
    /// - Easy to maintain and understand
    private func addMakeupItem(
        _ items: inout [BeautyItemInfo],
        name: String,
        icon: UIImage?,
        makeupName: String
    ) {
        items.append(
            BeautyItemInfo(
                name: name,
                icon: icon,
                value: beautyConfig.makeupIntensity * 100.0,  // 0.0~1.0 → 0~100
                isSelected: beautyConfig.makeupName == makeupName,
                valueRange: 0.0...100.0,
                // User drags slider: update intensity value (will sync update cache)
                onValueChanged: { [weak self] uiVal in
                    self?.beautyConfig.makeupIntensity = uiVal / 100.0  // 0~100 → 0.0~1.0
                },
                // User clicks to switch makeup
                onItemClick: { [weak self] itemInfo in
                    guard let self = self else { return }
                    
                    // 1. First get cached intensity value for this makeup (if user set it before)
                    let cachedIntensity = beautyConfig.getMakeupIntensityForTemplate(makeupName)
                    
                    // 2. Switch makeup (this will trigger SDK to set default intensity)
                    self.beautyConfig.makeupName = makeupName
                    
                    // 3. If cache has intensity value for this makeup (whether 0 or not), override SDK's default with cached value
                    //    This way when user switches back to previous makeup, it maintains previously set intensity value (including when user set to 0)
                    if let cached = cachedIntensity {
                        self.beautyConfig.makeupIntensity = cached
                    }
                    
                    // 4. Update UI displayed intensity value to ensure UI matches SDK state (convert to UI value)
                    itemInfo.value = self.beautyConfig.makeupIntensity * 100.0  // 0.0~1.0 → 0~100
                }
            )
        )
    }
}

File: Android/lib/src/main/java/cn/shengwang/beauty/ui/builder/MakeupPageBuilder.kt
package cn.shengwang.beauty.ui.builder

import cn.shengwang.beauty.R
import cn.shengwang.beauty.core.ShengwangBeautyManager
import cn.shengwang.beauty.core.MakeupNames
import cn.shengwang.beauty.ui.contract.IPageBuilder
import cn.shengwang.beauty.ui.model.BeautyPageInfo
import cn.shengwang.beauty.ui.model.BeautyItemInfo
import cn.shengwang.beauty.ui.model.BeautyItemType
import cn.shengwang.beauty.ui.model.BeautyModule

/**
 * 风格妆模块页面构建器
 * 负责构建风格妆模块的页面信息
 * 
 * 注意：此构建器为内部实现，不对外暴露
 */
internal class MakeupPageBuilder(
    private val beautyConfig: ShengwangBeautyManager.BeautyConfig
) : IPageBuilder {

    override fun buildPage(): BeautyPageInfo {
        val makeupItems = mutableListOf<BeautyItemInfo>()
        
        makeupItems.add(
            BeautyItemInfo(
                R.string.beauty_effect_none,
                R.drawable.beauty_ic_none,
                isSelected = beautyConfig.makeupName == null,
                showSlider = false,
                type = BeautyItemType.NONE,
                onItemClick = {
                    beautyConfig.makeupName = null
                }
            )
        )

        // 美妆选项
        // 学妹妆
        addMakeupItem(
            makeupItems,
            R.string.beauty_makeup_young,
            R.drawable.beauty_ic_makeup_young,
            MakeupNames.YOUNG
        )
        // 学姐妆
        addMakeupItem(
            makeupItems,
            R.string.beauty_makeup_mature,
            R.drawable.beauty_ic_makeup_mature,
            MakeupNames.MATURE
        )
        // 气质妆
        addMakeupItem(
            makeupItems,
            R.string.beauty_makeup_aura,
            R.drawable.beauty_ic_makeup_aura,
            MakeupNames.AURA
        )
        // 白皙妆
        addMakeupItem(
            makeupItems,
            R.string.beauty_makeup_natural,
            R.drawable.beauty_ic_makeup_natural,
            MakeupNames.NATURAL
        )
        // 优雅妆
        addMakeupItem(
            makeupItems,
            R.string.beauty_makeup_graceful,
            R.drawable.beauty_ic_makeup_graceful,
            MakeupNames.GRACEFUL
        )
        // 粉晕妆
        addMakeupItem(
            makeupItems,
            R.string.beauty_makeup_charm,
            R.drawable.beauty_ic_makeup_charm,
            MakeupNames.CHARM
        )
        // 俏皮妆
        addMakeupItem(
            makeupItems,
            R.string.beauty_makeup_perky,
            R.drawable.beauty_ic_makeup_perky,
            MakeupNames.PERKY
        )
        // 少女妆
        addMakeupItem(
            makeupItems,
            R.string.beauty_makeup_maiden,
            R.drawable.beauty_ic_makeup_maiden,
            MakeupNames.MAIDEN
        )
        // 深邃妆
        addMakeupItem(
            makeupItems,
            R.string.beauty_makeup_insight,
            R.drawable.beauty_ic_makeup_insight,
            MakeupNames.INSIGHT
        )
        // 氤氲妆
        addMakeupItem(
            makeupItems,
            R.string.beauty_makeup_misty,
            R.drawable.beauty_ic_makeup_misty,
            MakeupNames.MISTY
        )

        return BeautyPageInfo(
            R.string.beauty_group_makeup,
            makeupItems,
            type = BeautyModule.STYLE_MAKEUP
        )
    }

    /**
     * 添加美妆选项
     * 
     * 1. 初始化时：
     *    - 如果当前选中的是这个模板，使用 beautyConfig.makeupIntensity（会从缓存读取）
     *    - 如果当前选中的不是这个模板，从缓存读取该模板的强度值（如果没有则显示 0f）
     * 2. 用户拖动滑块时：更新当前模板的强度值（会同步更新缓存）
     * 3. 用户点击切换模板时：
     *    - 先切换模板（makeupName），这会触发 getter 从缓存读取该模板的强度值
     *    - 然后更新 UI 显示的强度值，确保 UI 与 SDK 状态一致
     * 
     * 这样设计的好处：
     * - 每个模板的强度值都会被缓存，切换回来时保持用户之前设置的强度
     * - API 层统一管理缓存，UI 层逻辑简单清晰
     * - 易于维护和理解
     */
    private fun addMakeupItem(
        items: MutableList<BeautyItemInfo>,
        nameRes: Int,
        iconRes: Int,
        makeupName: String
    ) {
        items.add(
            BeautyItemInfo(
                nameRes,
                iconRes,
                beautyConfig.makeupIntensity,
                isSelected = beautyConfig.makeupName == makeupName,
                valueRange = 0f..1.0f,
                // 用户拖动滑块时，更新强度值（会同步更新缓存）
                onValueChanged = { value ->
                    beautyConfig.makeupIntensity = value
                },
                // 用户点击切换模板时
                onItemClick = { itemInfo ->
                    // 1. 先获取缓存中该模板的强度值（如果用户之前设置过）
                    val cachedIntensity = beautyConfig.getMakeupIntensityForTemplate(makeupName)
                    
                    // 2. 切换模板（这会触发 SDK 设置默认强度）
                    beautyConfig.makeupName = makeupName
                    
                    // 3. 如果缓存中有该模板的强度值（无论是否为 0），用缓存的强度值覆盖 SDK 设置的默认值
                    //    这样用户切换回之前的模板时，会保持之前设置的强度值（包括用户设置为 0 的情况）
                    cachedIntensity?.let {
                        beautyConfig.makeupIntensity = it
                    }
                    
                    // 4. 更新 UI 显示的强度值，确保与 SDK 状态一致
                    itemInfo.value = beautyConfig.makeupIntensity
                }
            )
        )
    }
}
